More detailed resources usage

xXFiEsTaDeAmOnXx opened the pull request:
## Describe your changes
As requested in [issue #1847](https://github.com/end-4/dots-hyprland/issues/1847), this pull request adds additional CPU information to the `ResourcesPopup.qml` module. This includes the CPU frequency and the CPU temperature. 

A prieview:

<img width="2559" height="176" alt="image" src="https://github.com/user-attachments/assets/1d12bde8-a9ed-4842-a49d-fbdf37ccf019" />


## Is it ready? Questions/feedback needed?

The CPU info has been fully tested on both my desktop and laptop and works as expected.

We might consider adding GPU-related metrics for a more complete system overview.

reakjra commented:
I polished both AMD scripts:

`get_dgpuinfo.sh`
```diff
- | best_boot_vga=1
+ | best_boot_vga=0
```

<details><summary>get_igpuinfo.sh</summary>

```sh
# AMD iGPU
if ls /sys/class/drm/card*/device 1>/dev/null 2>&1; then
    echo "[AMD GPU - iGPU only]"

    card_path=""

    # prefer AMD with boot_vga=1 and no dedicated VRAM
    for d in /sys/class/drm/card*/device; do
        [[ -r "$d/vendor" ]] || continue
        if grep -qi "0x1002" "$d/vendor"; then
            boot=0; [[ -r "$d/boot_vga" ]] && boot=$(cat "$d/boot_vga")
            vtot=0
            if [[ -r "$d/mem_info_vis_vram_total" ]]; then
                vtot=$(cat "$d/mem_info_vis_vram_total")
            elif [[ -r "$d/mem_info_vram_total" ]]; then
                vtot=$(cat "$d/mem_info_vram_total")
            fi
            if [[ "$boot" == "1" ]] && (( vtot == 0 )); then
                card_path="$d"
                break
            fi
        fi
    done

    # fallback: with less VRAM
    if [[ -z "$card_path" ]]; then
        best_vram_total=-1
        for d in /sys/class/drm/card*/device; do
            [[ -r "$d/vendor" ]] || continue
            if grep -qi "0x1002" "$d/vendor"; then
                vtot=0
                if [[ -r "$d/mem_info_vis_vram_total" ]]; then
                    vtot=$(cat "$d/mem_info_vis_vram_total")
                elif [[ -r "$d/mem_info_vram_total" ]]; then
                    vtot=$(cat "$d/mem_info_vram_total")
                fi
                if (( best_vram_total < 0 )) || (( vtot < best_vram_total )); then
                    best_vram_total=$vtot
                    card_path="$d"
                fi
            fi
        done
    fi

    # manual override
    if [[ -n "${AMD_GPU_CARD:-}" && -d "/sys/class/drm/${AMD_GPU_CARD}/device" ]]; then
        card_path="/sys/class/drm/${AMD_GPU_CARD}/device"
    fi

    if [[ -n "$card_path" ]]; then
        [[ -r "$card_path/gpu_busy_percent" ]] && gpu_usage=$(cat "$card_path/gpu_busy_percent") || gpu_usage=0

        used_b=0; total_b=0
        if [[ -r "$card_path/mem_info_vis_vram_used" && -r "$card_path/mem_info_vis_vram_total" ]]; then
            used_b=$(cat "$card_path/mem_info_vis_vram_used")
            total_b=$(cat "$card_path/mem_info_vis_vram_total")
        elif [[ -r "$card_path/mem_info_vram_used" && -r "$card_path/mem_info_vram_total" ]]; then
            used_b=$(cat "$card_path/mem_info_vram_used")
            total_b=$(cat "$card_path/mem_info_vram_total")
        else
            vram_total_kib=$(grep MemTotal /proc/meminfo | awk '{print $2}')
            vram_available_kib=$(grep MemAvailable /proc/meminfo | awk '{print $2}')
            used_b=$(( (vram_total_kib - vram_available_kib) * 1024 ))
            total_b=$(( vram_total_kib * 1024 ))
        fi

        vram_used_gb=$(awk -v u="${used_b:-0}" 'BEGIN{printf "%.1f", u/1024/1024/1024}')
        vram_total_gb=$(awk -v t="${total_b:-0}" 'BEGIN{printf "%.1f", t/1024/1024/1024}')

        temperature=0; found=0
        for hm in "$card_path"/hwmon/hwmon*; do
            [[ -d "$hm" ]] || continue
            for lbl in "$hm"/temp*_label; do
                [[ -r "$lbl" ]] || continue
                if grep -qi "edge" "$lbl"; then
                    base="${lbl%_label}"
                    if [[ -r "${base}_input" ]]; then
                        temperature=$(awk '{printf "%.0f",$1/1000}' "${base}_input"); found=1; break
                    fi
                fi
            done
            [[ $found -eq 1 ]] && break
            if [[ $found -eq 0 ]]; then
                for lbl in "$hm"/temp*_label; do
                    [[ -r "$lbl" ]] || continue
                    if grep -qi "junction" "$lbl"; then
                        base="${lbl%_label}"
                        if [[ -r "${base}_input" ]]; then
                            temperature=$(awk '{printf "%.0f",$1/1000}' "${base}_input"); found=1; break
                        fi
                    fi
                done
                [[ $found -eq 1 ]] && break
            fi
            if [[ $found -eq 0 ]]; then
                for tin in "$hm"/temp*_input; do
                    [[ -r "$tin" ]] || continue
                    temperature=$(awk '{printf "%.0f",$1/1000}' "$tin"); found=1; break
                done
            fi
            [[ $found -eq 1 ]] && break
        done

        echo "  Usage : ${gpu_usage}%"
        echo "  VRAM  : ${vram_used_gb}/${vram_total_gb} GB"
        echo "  Temp  : ${temperature} °C"
        exit 0
    fi
fi
```
</details>

Now it should prioritize boot_vga=1 (which is iGPU, usually) or fallback on VRAM method but with less VRAM priority. While the dGPU script just puts best boot_vga=0. Might need to fix the iGPU output parsing

xXFiEsTaDeAmOnXx commented:
Alright, I included the changes.

xXFiEsTaDeAmOnXx commented:
Okay then the script reads the output correctly.

This may fix it.

The issue happens somewhere within this lines in `.config/quickshell/ii/services/ResourceUsage.qml`.

```qml
  Process {
    id: iGpuinfoProc
    command: ["bash", "-c", `${Directories.scriptPath}/gpu/get_igpuinfo.sh`.replace(/file:\/\//, "")]
    running: true

    stdout: StdioCollector {
      onStreamFinished:{
        iGpuAvailable =  this.text.indexOf("No GPU available") ==-1
        if(iGpuAvailable){
          iGpuUsage = this.text.match(/\sUsage\s:\s(\d+)/)?.[1] /  100 ?? 0
          const vramLine = this.text.match(/\sVRAM\s:\s(\d+(?:\.\d+)?)\/(\d+(?:\.\d+)?)\s*GB/)
          iGpuVramUsedGB = Number(vramLine?.[1] ?? 0)
          iGpuVramTotalGB = Number(vramLine?.[2] ?? 0)
          iGpuVramUsage = iGpuVramTotalGB > 0 ? (iGpuVramUsedGB / iGpuVramTotalGB) : 0;
          iGpuTempemperature = this.text.match(/\sTemp\s:\s(\d+)/)?.[1] ?? 0 
        }
       }
    }
  }

}
```

Can you try using `console.log()` within this section for example like this:

```qml

  iGpuUsage = this.text.match(/\sUsage\s:\s(\d+)/)?.[1] /  100 ?? 0
 console.log(iGpuUsage)
```
and see if the regex pattern does work correctly.

Unfortunately I can not do that as I only have Intel IGPU

xXFiEsTaDeAmOnXx commented:
> Okay then the script reads the output correctly.
> 
> This may fix it.
> 
> The issue happens somewhere within this lines in `.config/quickshell/ii/services/ResourceUsage.qml`.
> 
> ```qml
>   Process {
>     id: iGpuinfoProc
>     command: ["bash", "-c", `${Directories.scriptPath}/gpu/get_igpuinfo.sh`.replace(/file:\/\//, "")]
>     running: true
> 
>     stdout: StdioCollector {
>       onStreamFinished:{
>         iGpuAvailable =  this.text.indexOf("No GPU available") ==-1
>         if(iGpuAvailable){
>           iGpuUsage = this.text.match(/\sUsage\s:\s(\d+)/)?.[1] /  100 ?? 0
>           const vramLine = this.text.match(/\sVRAM\s:\s(\d+(?:\.\d+)?)\/(\d+(?:\.\d+)?)\s*GB/)
>           iGpuVramUsedGB = Number(vramLine?.[1] ?? 0)
>           iGpuVramTotalGB = Number(vramLine?.[2] ?? 0)
>           iGpuVramUsage = iGpuVramTotalGB > 0 ? (iGpuVramUsedGB / iGpuVramTotalGB) : 0;
>           iGpuTempemperature = this.text.match(/\sTemp\s:\s(\d+)/)?.[1] ?? 0 
>         }
>        }
>     }
>   }
> 
> }
> ```
> 
> Can you try using `console.log()` within this section for example like this:
> 
> ```qml
>   iGpuUsage = this.text.match(/\sUsage\s:\s(\d+)/)?.[1] /  100 ?? 0
>  console.log(iGpuUsage)
> ```
> 
> and see if the regex pattern does work correctly.
> 
> Unfortunately I can not do that as I only have Intel IGPU

Are you able to fix the regex within the qml to match the scripts output?

How does this this command to get the cpu temp look with amd cpu?
On intel it looks like this:
```bash
bash -c "paste <(cat /sys/class/thermal/thermal_zone*/type) <(cat /sys/class/thermal/thermal_zone*/temp)"

acpitz	16800
acpitz	27800
x86_pkg_temp	26000
```

Will this work on amd?

xXFiEsTaDeAmOnXx commented:
Yeah, you can add your code to this PR if you want.

LeDeliJelly commented:
<img width="781" height="165" alt="image" src="https://github.com/user-attachments/assets/95f23b02-1b19-4d05-946e-41dd9dd34bb3" />
CPU temp seems to work now.

Kupiuwu commented:
i use amd cpu and nvidia gpu (no igpu). Cpu temp works but gpu is not
<img width="420" height="110" alt="image" src="https://github.com/user-attachments/assets/266eea5e-b934-4836-a1ed-1c4f97faf4ef" />

reakjra commented:
@LeDeliJelly from what I've seen you own an AMD iGPU? Could you try this script? 

<details><summary>get_igpuinfo.sh</summary>

```sh
if ls /sys/class/drm/card*/device 1>/dev/null 2>&1; then
  echo "[AMD GPU - iGPU only]"

  card_path=""

  # override
  if [[ -n "${AMD_GPU_CARD:-}" && -d "/sys/class/drm/${AMD_GPU_CARD}/device" ]]; then
    card_path="/sys/class/drm/${AMD_GPU_CARD}/device"
  else
    best=""
    best_score=-1

    for d in /sys/class/drm/card*/device; do
      [[ -r "$d/vendor" ]] || continue
      grep -qi "0x1002" "$d/vendor" || continue

      # Dedicated VRAM total (if exists)
      vtot=0
      if [[ -r "$d/mem_info_vis_vram_total" ]]; then
        vtot=$(cat "$d/mem_info_vis_vram_total")
      elif [[ -r "$d/mem_info_vram_total" ]]; then
        vtot=$(cat "$d/mem_info_vram_total")
      fi

      # VRAM type (if exists)
      vtype=""; [[ -r "$d/vram_type" ]] && vtype=$(tr '[:upper:]' '[:lower:]' < "$d/vram_type")

      # GTT (system memory used by iGPU)
      gtt=0; [[ -r "$d/gtt_total" ]] && gtt=$(cat "$d/gtt_total")

      # Is there a connected display on this card? (nice hint)
      has_connected=0
      for con in /sys/class/drm/"${d##*/}"-*/status; do
        [[ -r "$con" ]] || continue
        if [[ "$(cat "$con")" == "connected" ]]; then
          has_connected=1; break
        fi
      done

      # Ddedicated VRAM must be 0 or vram_type must be "none" & GTT must be > 0
      is_igpu=0
      if { [[ "$vtot" -eq 0 ]] || [[ "$vtype" == "none" ]]; } && (( gtt > 0 )); then
        is_igpu=1
      fi
      (( is_igpu == 1 )) || continue

      # Score candidates to pick the "best" iGPU
      score=0
      [[ -r "$d/gpu_busy_percent" ]] && score=$((score+2))
      (( has_connected == 1 )) && score=$((score+1))
      if [[ -r "$d/boot_vga" && "$(cat "$d/boot_vga")" == "1" ]]; then
        score=$((score+1))
      fi

      if (( score > best_score )); then
        best="$d"; best_score=$score
      fi
    done

    card_path="$best"
  fi

  if [[ -z "${card_path}" ]]; then
    echo "No AMD iGPU found."
  else
    gpu_usage=0
    [[ -r "$card_path/gpu_busy_percent" ]] && gpu_usage=$(cat "$card_path/gpu_busy_percent")

    # vis_vram > vram > gtt > system fallback
    used_b=0; total_b=0
    if [[ -r "$card_path/mem_info_vis_vram_used" && -r "$card_path/mem_info_vis_vram_total" ]]; then
      used_b=$(cat "$card_path/mem_info_vis_vram_used")
      total_b=$(cat "$card_path/mem_info_vis_vram_total")
    elif [[ -r "$card_path/mem_info_vram_used" && -r "$card_path/mem_info_vram_total" ]]; then
      used_b=$(cat "$card_path/mem_info_vram_used")
      total_b=$(cat "$card_path/mem_info_vram_total")
    elif [[ -r "$card_path/gtt_used" && -r "$card_path/gtt_total" ]]; then
      used_b=$(cat "$card_path/gtt_used")
      total_b=$(cat "$card_path/gtt_total")
    else
      # Fallback: approximate with system ram (might be atrociously misleading)
      vram_total_kib=$(grep MemTotal /proc/meminfo | awk '{print $2}')
      vram_available_kib=$(grep MemAvailable /proc/meminfo | awk '{print $2}')
      used_b=$(( (vram_total_kib - vram_available_kib) * 1024 ))
      total_b=$(( vram_total_kib * 1024 ))
    fi

    vram_used_gb=$(awk -v u="${used_b:-0}" 'BEGIN{printf "%.1f", u/1024/1024/1024}')
    vram_total_gb=$(awk -v t="${total_b:-0}" 'BEGIN{printf "%.1f", t/1024/1024/1024}')

    # edge > junction > tctl > any temp*_input
    temperature=0; found=0
    for hm in "$card_path"/hwmon/hwmon*; do
      [[ -d "$hm" ]] || continue

      for key in edge junction Tctl; do
        for lbl in "$hm"/temp*_label; do
          [[ -r "$lbl" ]] || continue
          if grep -qi "$key" "$lbl"; then
            base="${lbl%_label}"
            if [[ -r "${base}_input" ]]; then
              temperature=$(awk '{printf "%.0f",$1/1000}' "${base}_input"); found=1; break
            fi
          fi
        done
        [[ $found -eq 1 ]] && break
      done

      if [[ $found -eq 0 ]]; then
        for tin in "$hm"/temp*_input; do
          [[ -r "$tin" ]] || continue
          temperature=$(awk '{printf "%.0f",$1/1000}' "$tin"); found=1; break
        done
      fi

      [[ $found -eq 1 ]] && break
    done

    echo "  Usage : ${gpu_usage} %"
    echo "  VRAM : ${vram_used_gb}/${vram_total_gb} GB"
    echo "  Temp : ${temperature} °C"
    exit 0
  fi
fi

```

</details>

reakjra commented:
> > @LeDeliJelly from what I've seen you own an AMD iGPU? Could you try this script?
> > get_igpuinfo.sh
> 
> <img alt="image" width="761" height="240" src="https://private-user-images.githubusercontent.com/92609665/486555243-55250de6-ecc8-4416-a9e7-2954a2af2e4c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDIxMTksIm5iZiI6MTc1NzM0MTgxOSwicGF0aCI6Ii85MjYwOTY2NS80ODY1NTUyNDMtNTUyNTBkZTYtZWNjOC00NDE2LWE5ZTctMjk1NGEyYWYyZTRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA4VDE0MzAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmOTg5YjRiOTdmZGQxMTZhMjY4ZTg2ZjNmZjBiMzdhMWFkYzQwYzMyNzUzM2I4OWM5MjEzOWNjM2I5M2RkZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._xr9Dnsy5C27TeFP2kP2ZPbcUdW_3PwmzqmEppXvZPE"> same popup....

can you try to run it manually and try the override?  

```sh
ls /sys/class/drm/
```
check every cardX listed and try them (swap X with the card number):
```sh
export AMD_GPU_CARD=cardX
```
and in the same shell run the script 

@LeDeliJelly

reakjra commented:
> Okay, great that it's working now. @reakjra I updated the AMD iGPU script according to your latest suggestion. Is that sufficient? It should work for all vendors now, right?

I suppose yes, it should work for all vendors. Tho I'm not sure someone tried with a intel dGPU? But shouldn't be really hard.




> I also have an amd dgpu I can slot into my framework laptop to test amd igpu+dgpu but I dunno when I'd get the chance to do that.

Whenever you have the chance it would be great to try out!

homebobhomebob commented:
End_4 has changed his dotfiles folder structure, i cant see the values of my temps, (show me NaN) may be it needs a review? Could you see if thats is true? or only me?

xXFiEsTaDeAmOnXx commented:
This looks great 👍.  Should I add you as contributer, so you can add it? 
I do not think it should take much time as to port it for the other graphics, as the script for doing so are already there.

reakjra commented:
Since I don't really know how to make a proper patch file and given it does not show the files that are missing (as far as I know), I'll just stick to the usual method :P so you can take a proper look too about what I meant.

<details><summary>@scripts/gpu/get_amd_gpuinfo.sh</summary>

```bash
#!/usr/bin/env bash
set -euo pipefail
LC_NUMERIC=C

CARD_PATH=""

# Sourcing for hyprland 'env VAR=val'
if [[ -f "$HOME/.config/hypr/custom/env.conf" ]]; then
  # converts 'env VAR=val' in 'export VAR=val' and eval
  eval "$(sed -n 's/^env /export /p' "$HOME/.config/hypr/custom/env.conf")"
fi 
# This is just to make it easier and comfier to override, or at least I think so. I guess? I conceived? I figured? I 

# override: AMD_GPU_CARD=cardX
if [[ -n "${AMD_GPU_CARD:-}" && -d "/sys/class/drm/${AMD_GPU_CARD}/device" ]]; then
  CARD_PATH="/sys/class/drm/${AMD_GPU_CARD}/device"
else
  best_vram_total=-1
  best_boot_vga=1
  for d in /sys/class/drm/card*/device; do
    [[ -r "$d/vendor" ]] || continue
    if grep -qi "0x1002" "$d/vendor"; then
      vtot=0
      [[ -r "$d/mem_info_vram_total" ]] && vtot=$(<"$d/mem_info_vram_total")
      boot=1
      [[ -r "$d/boot_vga" ]] && boot=$(<"$d/boot_vga")
      if (( vtot > best_vram_total )) || { (( vtot == best_vram_total )) && [[ "$boot" == "0" && "$best_boot_vga" != "0" ]]; }; then
        CARD_PATH="$d"
        best_vram_total=$vtot
        best_boot_vga=$boot
      fi
    fi
  done
fi

# emppty JSON
if [[ -z "$CARD_PATH" ]]; then
  echo '{}'
  exit 0
fi

# GPU name; not 100000000% reliable for accurate name
gpu_name="AMD GPU"
# BDF like 0000:03:00.0
bdf="$(basename "$(readlink -f "$CARD_PATH")")"

if command -v lspci >/dev/null 2>&1; then
  desc="$(LC_ALL=C lspci -s "$bdf" 2>/dev/null || true)"

  if [[ -n "$desc" ]]; then
    # tries to take between [ ... ] that contains "RADEON"
    if [[ "$desc" =~ \[([^\]]*Radeon[^\]]*)\] ]]; then
      bracket="${BASH_REMATCH[1]}"
      # Example bracket: "Radeon RX 9070/9070 XT/9070 GRE"

      # split per "/"
      IFS='/' read -r -a _parts <<< "$bracket"
      # trim
      parts=()
      for p in "${_parts[@]}"; do
        parts+=( "$(echo "$p" | sed -E 's/^[[:space:]]+|[[:space:]]+$//g')" )
      done

      # take prefix, remove series number
      # es: "Radeon RX 9070" => "Radeon RX"
      first="${parts[0]}"
      series_prefix="$(echo "$first" | sed -E 's/[[:space:]]*[0-9]+[[:space:]]*$//')"

      # choose better variable: XTX > XT > GRE > otherwise the first
      choose=""
      for p in "${parts[@]}";   do [[ "$p" =~ XTX ]] && choose="$p" && break; done
      if [[ -z "$choose" ]]; then
        for p in "${parts[@]}"; do [[ "$p" =~ XT  ]] && choose="$p" && break; done
      fi
      if [[ -z "$choose" ]]; then
        for p in "${parts[@]}"; do [[ "$p" =~ GRE ]] && choose="$p" && break; done
      fi
      if [[ -z "$choose" ]]; then
        choose="${parts[0]}"
      fi

      # if it does not contain "Radeon" already, reconstruct
      if [[ "$choose" != *Radeon* ]]; then
        gpu_name="$(echo "$series_prefix $(echo "$choose" | xargs)" | sed -E 's/[[:space:]]+/ /g')"
      else
        gpu_name="$choose"
      fi

    else
      # fallback: use everything after the ':', but remove "(rev ...)" and the vendor preamble
      # es: "VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Navi 48 [Radeon RX 9070 ...] (rev c0)"
      tmp="${desc#*: }"
      tmp="$(echo "$tmp" | sed -E 's/\(rev[^)]+\)//g' | sed -E 's/[[:space:]]+$//')"
      # if there's any bracket with "Radeon" in it, eat it.
      if [[ "$tmp" =~ \[([^\]]*Radeon[^\]]*)\] ]]; then
        gpu_name="${BASH_REMATCH[1]}"
      else
        gpu_name="AMD GPU"
      fi
    fi
  fi
fi

# GPU name override, just in case. (es: export AMD_GPU_NAME="Radeon RX 6950 XT")
if [[ -n "${AMD_GPU_NAME:-}" ]]; then
  gpu_name="$AMD_GPU_NAME"
fi

# escape for JSON
gpu_name_json=${gpu_name//\"/\\\"}



# usage %
if [[ -r "$CARD_PATH/gpu_busy_percent" ]]; then
  usage=$(<"$CARD_PATH/gpu_busy_percent")
else
  usage=0
fi

# VRAM
vram_used_b=0
vram_total_b=0
[[ -r "$CARD_PATH/mem_info_vram_used" ]]  && vram_used_b=$(<"$CARD_PATH/mem_info_vram_used")
[[ -r "$CARD_PATH/mem_info_vram_total" ]] && vram_total_b=$(<"$CARD_PATH/mem_info_vram_total")

if (( vram_total_b > 0 )); then
  vram_percent=$(( vram_used_b * 100 / vram_total_b ))
else
  vram_percent=0
fi

vram_used_gb=$(awk -v u="$vram_used_b" 'BEGIN{printf "%.1f", u/1024/1024/1024}')
vram_total_gb=$(awk -v t="$vram_total_b" 'BEGIN{printf "%.1f", t/1024/1024/1024}')

# Temps & fan RPM
edge="null"; junc="null"; mem="null"; rpm="null"

for hm in "$CARD_PATH"/hwmon/hwmon*; do
  [[ -d "$hm" ]] || continue
  shopt -s nullglob

  # temps with label
  for lbl in "$hm"/temp*_label; do
    base="${lbl%_label}"
    read -r labeltxt < "$lbl" || continue
    labeltxt=$(echo "$labeltxt" | tr '[:upper:]' '[:lower:]')
    if [[ -r "${base}_input" ]]; then
      val=$(awk '{printf "%.0f",$1/1000}' "${base}_input")
      case "$labeltxt" in
        *edge*)                  edge=$val ;;
        *junction*|*hotspot*|*junc*) junc=$val ;;
        *mem*|*vram*)            mem=$val ;;
      esac
    fi
  done

  # fan RPM (first valid)
  for fin in "$hm"/fan*_input; do
    if [[ -r "$fin" ]]; then
      cand=$(<"$fin")
      if [[ "$cand" =~ ^[0-9]+$ ]]; then   # 0 is valid: fans are off (Zero RPM funvtion)
        rpm=$cand
        break
      fi
    fi
  done
done

# final JSON
printf '{'
printf '"name": "%s", ' "$gpu_name_json"
printf '"usagePercent": %d, "vramUsedGB": %.1f, "vramTotalGB": %.1f, "vramPercent": %d, ' "$usage" "$vram_used_gb" "$vram_total_gb" "$vram_percent"
printf '"fanRpm": %s, ' "${rpm}"
printf '"tempEdgeC": %s, ' "${edge}"
printf '"tempJunctionC": %s, ' "${junc}"
printf '"tempMemC": %s' "${mem}"
printf '}\n'
```
</details>



<details><summary>@services/GpuUsage.qml</summary>

```qml
pragma Singleton
pragma ComponentBehavior: Bound

import qs.modules.common
import QtQuick
import Quickshell
import Quickshell.Io

Singleton {
    id: root
    property bool available: false
    property string gpuName: ""
    property int usagePercent: 0
    property int vramPercent: 0
    property real vramUsedGB: 0.0
    property real vramTotalGB: 0.0
    property int fanRpm: -1
    property int tempEdgeC: -1
    property int tempJunctionC: -1
    property int tempMemC: -1

    // History for graphing
    readonly property int historyLength: Config?.options.resources.historyLength ?? 60
    property list<real> gpuUsageHistory: []

    property string maxAvailableGpuString: gpuName || "GPU"

    readonly property string scriptCmd: "~/.config/quickshell/ii/scripts/gpu/get_amd_gpuinfo.sh"

    function updateGpuUsageHistory() {
        gpuUsageHistory = [...gpuUsageHistory, usagePercent / 100.0]
        if (gpuUsageHistory.length > historyLength) {
            gpuUsageHistory.shift()
        }
    }

    Process {
        id: proc
        stdout: StdioCollector {
            onStreamFinished: {
                try {
                    const obj = JSON.parse(this.text || "{}")
                    const has = obj && Object.keys(obj).length > 0
                    available = has
                    gpuName = (obj?.name ?? "")
                    usagePercent = Math.max(0, Math.min(100, obj?.usagePercent ?? 0))
                    vramPercent = Math.max(0, Math.min(100, obj?.vramPercent ?? 0))
                    vramUsedGB = Number(obj?.vramUsedGB ?? 0)
                    vramTotalGB = Number(obj?.vramTotalGB ?? 0)

                    if (obj?.fanRpm === null || obj?.fanRpm === undefined || isNaN(Number(obj?.fanRpm))) {
                        fanRpm = -1
                    } else {
                        fanRpm = Number(obj.fanRpm)
                    }

                    tempEdgeC = Number.isFinite(obj?.tempEdgeC) ? obj.tempEdgeC : -1
                    tempJunctionC = Number.isFinite(obj?.tempJunctionC) ? obj.tempJunctionC : -1
                    tempMemC = Number.isFinite(obj?.tempMemC) ? obj.tempMemC : -1

                    root.updateGpuUsageHistory()
                } catch (e) {
                    available = false
                    gpuName = ""
                    fanRpm = -1
                }
            }
        }
    }

    Timer {
        interval: 1
        running: true
        repeat: true
        onTriggered: {
            proc.exec({ command: ["bash", "-lc", scriptCmd] })
            interval = Config.options?.resources?.updateInterval ?? 3000
        }
    }
}

```

</details>


<details><summary>@modules/overlay/resources/resources.qml</summary>

```qml
pragma ComponentBehavior: Bound
import QtQuick
import QtQuick.Controls
import QtQuick.Layouts
import Quickshell
import Quickshell.Hyprland
import Qt5Compat.GraphicalEffects
import Qt.labs.synchronizer
import qs
import qs.services
import qs.modules.common
import qs.modules.common.widgets
import qs.modules.overlay

StyledOverlayWidget {
    id: root
    property list<var> resources: [
        {
            "icon": "planner_review",
            "name": Translation.tr("CPU"),
            "history": ResourceUsage.cpuUsageHistory,
            "maxAvailableString": ResourceUsage.maxAvailableCpuString
        },
        {
            "icon": "memory",
            "name": Translation.tr("RAM"),
            "history": ResourceUsage.memoryUsageHistory,
            "maxAvailableString": ResourceUsage.maxAvailableMemoryString
        },
        {
            "icon": "swap_horiz",
            "name": Translation.tr("Swap"),
            "history": ResourceUsage.swapUsageHistory,
            "maxAvailableString": ResourceUsage.maxAvailableSwapString
        },
        {
            "icon": "empty_dashboard",
            "name": Translation.tr("GPU"),
            "history": GpuUsage.gpuUsageHistory,
            "maxAvailableString": GpuUsage.maxAvailableGpuString,
            "type": "gpu"
        },
    ]

    contentItem: Rectangle {
        id: contentItem
        anchors.centerIn: parent
        color: Appearance.m3colors.m3surfaceContainer
        radius: root.contentRadius
        property real padding: 4
        implicitWidth: 350
        implicitHeight: root.resources[tabBar.currentIndex]?.type === "gpu" ? 300 : 200
        // implicitHeight: contentColumn.implicitHeight + padding * 2
        ColumnLayout {
            id: contentColumn
            anchors {
                fill: parent
                margins: parent.padding
            }
            spacing: 8

            SecondaryTabBar {
                id: tabBar

                currentIndex: Persistent.states.overlay.resources.tabIndex
                onCurrentIndexChanged: {
                    Persistent.states.overlay.resources.tabIndex = tabBar.currentIndex;
                }

                Repeater {
                    model: root.resources.length
                    delegate: SecondaryTabButton {
                        required property int index
                        property var modelData: root.resources[index]
                        buttonIcon: modelData.icon
                        buttonText: modelData.name
                    }
                }
            }

            ResourceSummary {
                Layout.margins: 8
                history: root.resources[tabBar.currentIndex]?.history ?? []
                maxAvailableString: root.resources[tabBar.currentIndex]?.maxAvailableString ?? "--"
            }
            // GPU detailed info
            GpuDetails {
                Layout.margins: 8
                Layout.fillWidth: true
                visible: root.resources[tabBar.currentIndex]?.type === "gpu"
            }
        }
    }

    component ResourceSummary: RowLayout {
        id: resourceSummary
        required property list<real> history
        required property string maxAvailableString
        Layout.fillWidth: true
        Layout.fillHeight: true
        spacing: 12

        ColumnLayout {
            spacing: 2
            StyledText {
                text: (resourceSummary.history[resourceSummary.history.length - 1] * 100).toFixed(1) + "%"
                font {
                    family: Appearance.font.family.numbers
                    variableAxes: Appearance.font.variableAxes.numbers
                    pixelSize: Appearance.font.pixelSize.huge
                }
            }
            StyledText {
                text: Translation.tr("of %1").arg(resourceSummary.maxAvailableString)
                font {
                    // family: Appearance.font.family.numbers
                    // variableAxes: Appearance.font.variableAxes.numbers
                    pixelSize: Appearance.font.pixelSize.smallie
                }
                color: Appearance.colors.colSubtext
            }
            Item {
                Layout.fillHeight: true
            }
        }
        Rectangle {
            id: graphBg
            Layout.fillWidth: true
            Layout.fillHeight: true
            radius: Appearance.rounding.small
            color: Appearance.colors.colSecondaryContainer
            layer.enabled: true
            layer.effect: OpacityMask {
                maskSource: Rectangle {
                    width: graphBg.width
                    height: graphBg.height
                    radius: graphBg.radius
                }
            }
            Graph {
                anchors.fill: parent
                values: root.resources[tabBar.currentIndex]?.history ?? []
                points: ResourceUsage.historyLength
                alignment: Graph.Alignment.Right
            }
        }
    }
    component GpuDetails: ColumnLayout {
        spacing: 6

        GpuInfoRow {
            iconName: "bolt"
            label: "Usage"
            value: GpuUsage.usagePercent + "%"
        }

        GpuInfoRow {
            iconName: "memory_alt"
            label: "VRAM"
            value: GpuUsage.vramUsedGB.toFixed(1) + "/" + GpuUsage.vramTotalGB.toFixed(1) + " GB"
        }

        GpuInfoRow {
            iconName: "air"
            label: "Fan"
            value: (GpuUsage.fanRpm >= 0) ? GpuUsage.fanRpm + " RPM" : "—"
        }

        GpuInfoRow {
            iconName: "thermometer"
            label: "Edge"
            value: (GpuUsage.tempEdgeC >= 0) ? GpuUsage.tempEdgeC + " °C" : "—"
        }

        GpuInfoRow {
            iconName: "thermometer"
            label: "Junction"
            value: (GpuUsage.tempJunctionC >= 0) ? GpuUsage.tempJunctionC + " °C" : "—"
        }

        GpuInfoRow {
            iconName: "thermometer"
            label: "Memory"
            value: (GpuUsage.tempMemC >= 0) ? GpuUsage.tempMemC + " °C" : "—"
        }
    }

    component GpuInfoRow: RowLayout {
        required property string iconName
        required property string label
        required property string value
        Layout.fillWidth: true
        spacing: 6

        MaterialSymbol {
            text: iconName
            iconSize: Appearance.font.pixelSize.normal
            color: Appearance.colors.colOnSurfaceVariant
        }

        StyledText {
            text: label + ":"
            color: Appearance.colors.colOnSurfaceVariant
            font.pixelSize: Appearance.font.pixelSize.small
        }

        StyledText {
            Layout.fillWidth: true
            horizontalAlignment: Text.AlignRight
            text: value
            color: Appearance.colors.colOnSurfaceVariant
            font {
                pixelSize: Appearance.font.pixelSize.small
                family: Appearance.font.family.numbers
            }
        }
    }
    }

```

</details>


> The AMD script is mostly the same, it's just cleaner and it'll try to catch the GPU name, but there's an override too (still pointing to the DGPU tho).

xXFiEsTaDeAmOnXx commented:
Okay I added new fileds for the gpu:

<img width="365" height="337" alt="image" src="https://github.com/user-attachments/assets/85376eb0-5046-497b-8356-2353c9cd9333" />


The AMD calculation for that is still missing. I cannot add all fileds as some of them or only available for nvidia and some only for amd. I added some more which should be available for both. I also added the new service you mentioned.

end-4 commented on dots/.config/quickshell/ii/modules/overlay/resources/Resources.qml:
```diff
@@ -13,27 +13,148 @@ import qs.modules.common.widgets
-    property list<var> resources: [
+    function formatKB(kb) {
+        return (kb / (1024 * 1024)).toFixed(1) + " GB";
+    }
```
use `ResourceUsage.kbToGbString` instead of copying this over

end-4 commented on dots/.config/quickshell/ii/scripts/gpu/get_dgpuinfo.sh:
```diff
@@ -0,0 +1,171 @@
+            [[ $found -eq 1 ]] && break
+        done
+        
+        # TODO Add AMD calcualtion
```
typo

end-4 commented on dots/.config/quickshell/ii/scripts/gpu/get_dgpuinfo.sh:
I'm highly skeptical of this because the power state file seems kind of arbitrary and seems specific, but whatever

reakjra commented:
@end-4 I would consider not minding the current scripts, I'm rewriting them properly and with JSON outputs which are more reliable.

end-4 commented:
These lines at the bottom are not necessary. The graph is available for a reason.

<img width="399" height="376" alt="image" src="https://github.com/user-attachments/assets/bed15c24-8195-4c43-8814-3381ba1c4f27" />

end-4 commented:
With this PR the CPU always spikes to 5% (instead of hovering around 1%)... That will have to be fixed or all these measurements must be toggleable, off by default

xXFiEsTaDeAmOnXx commented:
Okay, I will  also look  into this.

reakjra commented:
As I said, since commenting them here would be atrocious, maybe I can give you the edited files through google drive or whatever file sharing site exists and you test it in a clean config, or finding another way I can send the updates codes in a better way without files

reakjra commented:
I think there are a lot of things left behind. The scripts are not implemented, the Resources.qml in the overlay uses things that are missing or placed incorrectly in the Config.qml. In the files I've sent there are fixes for the typos such as 'Freqency" and "Tempemperature" too

reakjra commented:
technically I can create a new branch and then PR this branch, right? Since I can now work in your repo. I'm not sure tho. 

Well I'll try, I'll delete the branch if I'm wrong

xXFiEsTaDeAmOnXx commented:
Sure, you can do that.

sclewis900 commented:
> I played a bit with this branch, and it _almost_ works. There are some small issues with the way data is displayed in the resources tab (an extra parenthesis for CPU load, for example), as well as with the iGPU/dGPU shell scripts (eg. some parentheses not being properly escaped in regex). And, of course, the merge conflict.
> 
> You can enable the GPU data by manually editing the file `~/.config/illogical-impulse/config.json`. It should have some GPU-related sections in `resources`, as well as in `bar.resources`.

This is what I see in bar.resources:

"resources": {
            "alwaysShowCpu": true,
            "alwaysShowSwap": true,
            "cpuWarningThreshold": 90,
            "memoryWarningThreshold": 95,
            "swapWarningThreshold": 85
            

And this is what I see in resources:

"resources": {
        "historyLength": 60,
        "updateInterval": 3000
        

I tried adding "alwaysShowGpu":true, and nothing shows

sclewis900 commented:
this is the location, in case there is another Config.qml

/home/ven/.config/quickshell/ii/modules/common/Config.qml
